File: src/main/java/net/softbell/bsh/dto/view/general/MonitorCardItemDto.java
package net.softbell.bsh.dto.view.general;

import lombok.Getter;
import lombok.Setter;
import net.softbell.bsh.domain.entity.NodeItem;
import net.softbell.bsh.domain.entity.NodeItemHistory;

/**
 * @Author : Bell(bell@softbell.net)
 * @Description : 모니터뷰 카드 아이템 정보 DTO
 */
@Getter
@Setter
public class MonitorCardItemDto
{
	private String alias;
	private Double lastStatus;
	private Integer widthPercent;
	private Boolean isDigital;
	
	public MonitorCardItemDto(NodeItem entity, NodeItemHistory lastHistory)
	{
		// Exception
		if (entity == null)
			return;
		
		// Convert
		this.alias = entity.getAlias();
		this.lastStatus = lastHistory.getItemStatus();
		switch (entity.getItemMode())
		{
			case DIGITAL:
				isDigital = true;
				if (this.lastStatus == 0)
					this.widthPercent = 0;
				else
					this.widthPercent = 100;
				
				break;
				
			case ANALOG:
				isDigital = false;
				switch (entity.getItemType())
				{
					case SENSOR_HUMIDITY:
						this.widthPercent = this.lastStatus.intValue(); // 습도는 0~100%
						break;
						
					case SENSOR_TEMPERATURE:
						this.widthPercent = (int) (this.lastStatus + 20); // 온도는 -20~40
						break;
						
					case READER_RFID:
						if (this.lastStatus != 0)
							this.widthPercent = 100; // RFID는 탐지 여부로 판정
						else
							this.widthPercent = 0;
						break;
						
					default:
						this.widthPercent = (int) (this.lastStatus / 10);
				}
					
				break;
				
			default:
				this.widthPercent = 0;
				
				break;
		}
	}
}

File: src/main/kotlin/net/softbell/bsh/dto/view/general/MonitorCardItemDto.kt
package net.softbell.bsh.dto.view.general

import net.softbell.bsh.domain.ItemModeRule
import net.softbell.bsh.domain.ItemTypeRule
import net.softbell.bsh.domain.entity.NodeItem
import net.softbell.bsh.domain.entity.NodeItemHistory

/**
 * @author : Bell(bell@softbell.net)
 * @description : 모니터뷰 카드 아이템 정보 DTO
 */
class MonitorCardItemDto(entity: NodeItem, lastHistory: NodeItemHistory) {
    val alias: String = entity.alias
    val lastStatus: Double = lastHistory.itemStatus
    val widthPercent: Int
    val digital: Boolean

    init {
        // Convert
        when (entity.itemMode) {
            ItemModeRule.DIGITAL -> {
                digital = true
                widthPercent = if (lastStatus == 0.0)
                    0
                else
                    100
            }
            ItemModeRule.ANALOG -> {
                digital = false
                widthPercent = when (entity.itemType) {
                    ItemTypeRule.SENSOR_HUMIDITY -> lastStatus.toInt() // 습도는 0~100%
                    ItemTypeRule.SENSOR_TEMPERATURE -> (lastStatus + 20).toInt() // 온도는 -20~40
                    ItemTypeRule.READER_RFID -> if (lastStatus != 0.0)
                        100 // RFID는 탐지 여부로 판정
                    else
                        0
                    else -> (lastStatus / 10).toInt()
                }
            }
            else -> {
                digital = false
                widthPercent = 0
            }
        }
    }
}